posts/lint-rcpp/index.qmd

```diff
@@ -0,0 +1,168 @@
+The latter case is natural as they are released by different projects: _clang-format_ by LLVM (despite specifying `-style=google`), and _cpplint_ derives from Google's work.
+
+### Linter-enforced paradigms
+
+An example of a linter-enforced paradigm is that _cpplint_ warns against passing function arguments by reference, and prefers for these to be passed as pointers, or as constant references (`const int &value`).
+
+```c++
+int some_function(int &value) { /* some operations on value */ return value;}
+```
+
+Passing the argument as a `const` reference would not serve the needs of this function, and passing by value is a valid strategy when we don't want to get into the details of using pointers.
+(Note that this is typically an issue when large objects such as custom classes or structs are passed to a function multiple times.)
+
+### Linter-styler disagreement
+
+One example of linter-styler disagreement is the use of `BOOST_FOREACH` as an alternative to `for` loops.
+_clang-format_ will insist on adding two spaces before the opening bracket: `BOOST_FOREACH  ()`.
+_cpplint_ will insist on removing one space.
+_cpplint_ and _clang-format_ also disagree on the order of header inclusions, especially when both local and system headers are included.
+
+### Disabling checks on code chunks
+
+Either of these cases could require disabling checks on some part of the code.
+It is possible to turn off linting using _cpplint_ at particular lines using the comment `// NOLINT`.
+Multiple lines can be protected from linting as well.
+
+```c++
+// NOLINTBEGIN
+<some C++ code here>
+// NOLINTEND
+```
+
+Alternatively, _clang-format_ can be instructed to ignore chunks of code using comment messages too.
+
+```c++
+// clang-format off
+<some C++ code here>
+// clang-format on
+```
+
+## Linter options for future packages
+
+{finalsize} is a relatively simple {Rcpp} package, with no C/C++ headers, and no C++ tests.
+However, future Epiverse packages could be more similar to {fluEvidenceSynthesis}, and will have header files, and could also have [C++ unit tests via the _catch_ framework](https://github.com/catchorg/Catch2).
+
+_cpplint_ will demand that all local headers be prefixed with their directory (`src/`), but this would cause the code to break as {Rcpp} looks for a subdirectory called `src/src/`.
+This can be turned off by passing the filter option `--filter="-build/include_subdir"` to _cpplint_. 
+Alternatively, we can place headers in a subdirectory such as `inst/include`.
+
+Both _cpplint_ and _cppcheck_ can be instructed to ignore C++ test files using the _catch_ testing framework provided by {testthat}.
+This prevents errors due to the specialised syntax provided by {testthat} in `testthat.h`, such as `context`.
+
+```sh
+# for cpplint, add an extra exclude statement
+cpplint ... --exclude="src/test*.cpp" src/*.cpp
+
+# for cppcheck, suppress checks on test files
+cppcheck ... --suppress=*:src/test_*.cpp src
```

Comment:
```suggestion
# for cpplint, add an extra exclude statement
cpplint <...> --exclude="src/test*.cpp" src/*.cpp

# for cppcheck, suppress checks on test files
cppcheck <...> --suppress=*:src/test_*.cpp src
```

This may be less confusing because R users are used to `...` being a valid instruction`

Reply:
Fixed in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/7/commits/c2da24228d857320605ca609753a13f71441aaa3